### PR TITLE
[PodSecurity] Aggregate identical warnings for multiple pods in a namespace

### LINKED ISSUE #103213: [PodSecurity] Aggregate identical warnings for multiple pods in a namespace

### PR DESCRIPTION
**Pull request by njuptlzf**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
part-4 of #103561
#### Which issue(s) this PR fixes:

Fixes #103213

#### Special notes for your reviewer:
1. Use `map[string][]string` to storage the podname of `the same reason`
2. Convert unordered  map `podsWarings` to []string `warnings`, the order of the checkers that `output` warnings each time is not guaranteed.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/2579-psp-replacement
```

/sig auth security

### CODE REVIEW

**Comment by njuptlzf** [comment]
Wait for #103210 to merge first

/hold

**Comment by njuptlzf** [comment]
before this pr:
```shell
# kubectl label ns kube-system pod-security.kubernetes.io/enforce=baseline
Warning: calico-node-x52m7: host namespaces, hostPath volumes, privileged == true
Warning: kube-scheduler-192.168.122.1: host namespaces, hostPath volumes
Warning: calico-kube-controllers-6778878bb8-gmzwx: host namespaces
Warning: kube-controller-manager-192.168.122.1: host namespaces, hostPath volumes
Warning: kube-proxy-192.168.122.1: host namespaces, hostPath volumes, privileged == true
namespace/kube-system labeled
```

after this pr:
```shell
# kubectl label ns kube-system pod-security.kubernetes.io/enforce=baseline
Warning: hostPath volumes: calico-node-x52m7, kube-controller-manager-192.168.122.1, kube-scheduler-192.168.122.1, kube-proxy-192.168.122.1
Warning: privileged == true: calico-node-x52m7, kube-proxy-192.168.122.1
Warning: host namespaces: calico-kube-controllers-6778878bb8-gmzwx, calico-node-x52m7, kube-controller-manager-192.168.122.1, kube-scheduler-192.168.122.1, kube-proxy-192.168.122.1
namespace/kube-system labeled
```

**Comment by njuptlzf** [comment]
before this pr:
```shell
# kubectl label ns kube-system pod-security.kubernetes.io/enforce=baseline
Warning: calico-node-x52m7: host namespaces, hostPath volumes, privileged == true
Warning: kube-scheduler-192.168.122.1: host namespaces, hostPath volumes
Warning: calico-kube-controllers-6778878bb8-gmzwx: host namespaces
Warning: kube-controller-manager-192.168.122.1: host namespaces, hostPath volumes
Warning: kube-proxy-192.168.122.1: host namespaces, hostPath volumes, privileged == true
namespace/kube-system labeled
```

after this pr:
```shell
# kubectl label ns kube-system pod-security.kubernetes.io/enforce=baseline
Warning: host namespaces: calico-node-x52m7, kube-proxy-192.168.122.1, kube-controller-manager-192.168.122.1, calico-kube-controllers-6778878bb8-gmzwx, kube-scheduler-192.168.122.1
Warning: hostPath volumes: calico-node-x52m7, kube-proxy-192.168.122.1, kube-controller-manager-192.168.122.1, kube-scheduler-192.168.122.1
Warning: privileged == true: calico-node-x52m7, kube-proxy-192.168.122.1
namespace/kube-system labeled
```

File: staging/src/k8s.io/pod-security-admission/admission/admission.go [hunk 1]
```diff
@@ -437,21 +443,44 @@ func (a *Admission) EvaluatePodsInNamespace(ctx context.Context, namespace strin
 	for i, pod := range pods {
 		// short-circuit on exempt runtimeclass
 		if a.exemptRuntimeClass(pod.Spec.RuntimeClassName) {
+			if i == len(pods)-1 {
+				warnings = aggregateWarnings(podsWarings, warnings)
+			}
 			continue
 		}
-		r := policy.AggregateCheckResults(a.Evaluator.EvaluatePod(enforce, &pod.ObjectMeta, &pod.Spec))
+		r, reasons := policy.AggregateCheckResults(a.Evaluator.EvaluatePod(enforce, &pod.ObjectMeta, &pod.Spec))
```
Comment (pacoxu) [inline staging/src/k8s.io/pod-security-admission/admission/admission.go:451]: `reasons` is part of `r`.
```
reasons := r.ForbiddenReasons
```
A little confused about the func refactor of `AggregateCheckResults`.

Comment (njuptlzf) [reply staging/src/k8s.io/pod-security-admission/admission/admission.go:451]: Thanks for pointing out, there is indeed no need to refactor `AggregateCheckResults`.

**Comment by njuptlzf** [comment]
According to this [comment](https://github.com/kubernetes/kubernetes/issues/103213#issuecomment-887593049)

before this pr 
```shell
# kubectl label ns kube-system pod-security.kubernetes.io/enforce=baseline
Warning: kube-scheduler-192.168.122.1: host namespaces, hostPath volumes
Warning: kubia-deploy-674ddfb57f-vh97k: hostPath volumes
Warning: kube-controller-manager-192.168.122.1: host namespaces, hostPath volumes
Warning: kube-proxy-192.168.122.1: host namespaces, hostPath volumes, privileged
Warning: calico-node-x52m7: host namespaces, hostPath volumes, privileged
Warning: kubia-deploy-674ddfb57f-4sxcz: hostPath volumes
Warning: kubia-deploy-674ddfb57f-kxtbt: hostPath volumes
Warning: calico-kube-controllers-6778878bb8-gmzwx: host namespaces
Warning: metrics-server-5f8996465b-4gzxq: hostPath volumes
```
after this pr
```shell
# kubectl label ns kube-system pod-security.kubernetes.io/enforce=baseline
Warning: kubia-deploy-674ddfb57f-kxtbt (and 3 other pods): hostPath volumes
Warning: calico-kube-controllers-6778878bb8-gmzwx: host namespaces
Warning: kube-controller-manager-192.168.122.1 (and 1 other pod): host namespaces, hostPath volumes
Warning: calico-node-x52m7 (and 1 other pod): host namespaces, hostPath volumes, privileged
```

**Comment by njuptlzf** [comment]
Sorry, I just saw it now. Unfortunately I am participating in closed development, thank you @liggitt for achieving it.

**Pull request closed without merging**